ukvm/ukvm_module_gdb.c

```diff
@@ -132,195 +41,164 @@
+    return -1;
+}
 
-static void wait_for_connect(int portn)
+/*
+ * Converts the memory pointed to by mem into an hex string in buf.
+ * Returns a pointer to the last char put in buf (null).
+ */
+static char *mem2hex(char *mem, char *buf, int count)
 {
-    struct sockaddr_in sockaddr;
-    socklen_t sockaddr_len;
-    struct protoent *protoent;
-    int r;
-    int opt;
+    int i;
+    uint8_t ch;
+
+    for (i = 0; i < count; i++) {
+        ch = *mem++;
+        *buf++ = hexchars[ch >> 4];
+        *buf++ = hexchars[ch % 16];
+    }
+    *buf = 0;
+    return buf;
+}
 
-    printf("GDB trying to get a connection at port %d\n", portn);
+/*
+ * Converts the hex string in buf into binary in mem.
+ * Returns a pointer to the character AFTER the last byte written.
+ */
+static char *hex2mem(char *buf, char *mem, int count)
+{
+    int i;
+    uint8_t ch;
+
+    for (i = 0; i < count; i++) {
+        ch = hex(*buf++) << 4;
+        ch = ch + hex(*buf++);
+        *mem++ = ch;
+    }
+    return mem;
+}
 
-    listen_socket_fd = socket(PF_INET, SOCK_STREAM, 0);
-    assert(listen_socket_fd != -1);
+static int wait_for_connect()
+{
+    int listen_socket_fd;
+    struct sockaddr_in server_addr, client_addr;
+    struct protoent *protoent;
+    struct in_addr ip_addr;
+    socklen_t len;
+    int opt;
 
-    /* Allow rapid reuse of this port */
+    listen_socket_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (listen_socket_fd == -1) {
+        perror("GDB: Could not create socket");
+        return -1;
+    }
+    
     opt = 1;
-    r = setsockopt(listen_socket_fd, SOL_SOCKET, SO_REUSEADDR, &opt,
-                   sizeof(opt));
-    if (r == -1)
-        perror("setsockopt(SO_REUSEADDR) failed");
-
-    memset(&sockaddr, '\000', sizeof(sockaddr));
-    sockaddr.sin_family = AF_INET;
-    sockaddr.sin_port = htons(portn);
-    sockaddr.sin_addr.s_addr = htonl(INADDR_ANY);
-
-    r = bind(listen_socket_fd, (struct sockaddr *) &sockaddr,
-             sizeof(sockaddr));
-    if (r == -1)
-        perror("Failed to bind socket");
-
-    r = listen(listen_socket_fd, 0);
-    if (r == -1)
-        perror("Failed to listen on socket");
-
-    sockaddr_len = sizeof(sockaddr);
-    socket_fd = accept(listen_socket_fd, (struct sockaddr *) &sockaddr,
-                       &sockaddr_len);
-    if (socket_fd == -1)
-        perror("Failed to accept on socket");
+    if (setsockopt(listen_socket_fd, SOL_SOCKET, SO_REUSEADDR, &opt,
+                   sizeof(opt)) < 0)
+        perror("GDB: setsockopt(SO_REUSEADDR) failed");
+
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+    server_addr.sin_port = htons(portno);
+     
+    if (bind(listen_socket_fd, (struct sockaddr *)&server_addr,
+             sizeof(server_addr)) < 0) {
+        perror("GDB: bind failed");
+        return -1;
+    }
+     
+    listen(listen_socket_fd , 0);
+     
+    puts("GDB: Waiting for a debugger...");
```

Comment:
Use `warnx()` so that this is prefixed with the `ukvm` binary and goes to standard error, not standard output (which may be redirecting the guest console). Also, the message should mention the port it's expecting the connection on. Maybe suggest to the user the command to use? (`target remote :port`)